File: LeetCode/Kai/61RotateList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if(head == null) return head;  //corner case
        
        //create a cycle
        int count = 0;
        ListNode node = head;
        while(node.next != null) {
            node = node.next;
            count++;
        }
        node.next = head;
        
        //break cycle from new head
        for(int i=0; i<count-k%(count+1); i++) {
            head = head.next;
        }
        ListNode newHead= head.next;
        head.next = null;
        
        return newHead;
    }
    
}

File: LeetCode/Peng/61RotateList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (!head) return NULL;
        
        ListNode* current = head;
        int length = 0;
        while (current) {
            length++;
            current = current->next;
        }
        
        int tempK = k % length;
        current = head;
        ListNode* end = head;
        while (tempK) {
            end = end->next;
            tempK--;
        }
        while (end->next) {
            current = current->next;
            end = end->next;
        }
        end->next = head;
        head = current->next;
        current->next = NULL;
        return head;
    }
};